contract/src/contract.rs

```diff
@@ -95,24 +177,130 @@ pub fn update_config(
     Ok(Response::new().add_attribute("action", "update_config"))
 }
 
-pub fn burn(deps: DepsMut, env: Env, info: MessageInfo) -> Result<Response, ContractError> {
+/// Burns native tokens or CW20 tokens based on the contract's [Config] setup during instantiate
+/// If the contract is setup to use CW20 input tokens, the CW20 token address must be set in the config
+/// If the contract is setup to use native input tokens, the CW20 token address must not be set in the config
+/// If the contract is setup to mint native ASH tokens on burn, config.use_cw20 must be false
+/// If the contract is setup to mint CW20 ASH tokens on burn, config.use_cw20 must be true
+///
+/// In either case 3 main actions happen assuming good data
+/// 1. A burn message of the native input token (or cw20 if setup right)
+/// 2. A mint message of the native ASH token (or cw20 if setup right)
+/// 3. A transfer message of the native ASH token (or cw20 if setup right)
+pub fn burn(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    cw20_msg: Option<Cw20ReceiveMsg>,
+) -> Result<Response, ContractError> {
     //check the only one token is sent
-    if info.funds.len() != 1 {
-        return Err(ContractError::IncorrectTokenQuantity {});
-    }
-
-    //Only burn whale tokens
-    if info.funds[0].denom != "uwhale" {
-        return Err(ContractError::IncorrectToken {});
-    }
     let config: Config = CONFIG.load(deps.storage)?;
+
     let mut messages: Vec<CosmosMsg> = vec![];
-    let amount = info.funds[0].amount;
 
-    //burn the whale sent
-    messages.push(CosmosMsg::from(BankMsg::Burn {
-        amount: info.funds.clone(),
-    }));
+    match config.burn_cw20_addr {
+        Some(burn_cw20_addr) => {
+            // If the BurnCw20_addr was set, we need to not check the 'funds'
+            // We can check the input amount from the cw20_msg
+            // If the BurnCw20_addr is set, we need to check the cw20_msg
+            if cw20_msg.is_none() {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            let cw20_msg = cw20_msg.clone().unwrap();
+
+            if cw20_msg.amount.is_zero() {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            messages.push(build_burn_cw20_token_msg(
+                burn_cw20_addr.to_string(),
+                cw20_msg.amount,
+            )?);
+        }
+        None => {
+            if info.funds.len() != 1 {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            //Only burn whale tokens
+            if info.funds[0].denom != "uwhale" {
```

Comment:
We would like this denom to be variable in the future, on chihuahua for instance we will burn `uhuahua`

Reply:
Ooof great catch man, I need to change this as technically it works for all native but this will kill it 
Will make change + add test